Add Core Image accelerated path for SourceAlpha

nullhook opened the pull request:
#### 4a6741a21b543ab21819664808db58fda74fc444
<pre>
Add Core Image accelerated path for SourceAlpha
<a href="https://bugs.webkit.org/show_bug.cgi?id=297018">https://bugs.webkit.org/show_bug.cgi?id=297018</a>
<a href="https://rdar.apple.com/problem/157696906">rdar://problem/157696906</a>

Reviewed by Said Abou-Hallawa.

Implement accelerated rendering for the SourceAlpha
using Core Image&apos;s CIColorMatrix

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/coreimage/SourceAlphaCoreImageApplier.h: Added.
* Source/WebCore/platform/graphics/coreimage/SourceAlphaCoreImageApplier.mm: Added.
(WebCore::SourceAlphaCoreImageApplier::apply const):
* Source/WebCore/platform/graphics/filters/SourceAlpha.cpp:
(WebCore::SourceAlpha::supportedFilterRenderingModes const):
(WebCore::SourceAlpha::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/SourceAlpha.h:
* Source/WebCore/platform/mac/VideoPresentationInterfaceMac.mm:

Canonical link: <a href="https://commits.webkit.org/298317@main">https://commits.webkit.org/298317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40111390b0ec712c8c476ddadc3a3bd5d77bf6cd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115124 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34832 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25321 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121238 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65755 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a37cb722-c2fc-4288-ad1d-5e48c49f02bf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117013 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35484 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43398 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87478 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9ea3dae8-159f-43db-99b5-f24d7ad78b7a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118072 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28285 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103349 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67875 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27453 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21469 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64890 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97669 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21586 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124424 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42087 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31483 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96275 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42458 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99539 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96061 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24442 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41268 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19107 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41963 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47504 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41498 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44819 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43231 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebCore/platform/graphics/coreimage/SourceAlphaCoreImageApplier.mm:
```diff
@@ -0,0 +1,69 @@
+    ASSERT(inputs.size() == 1);
+    Ref input = inputs[0].get();
+
+    auto inputImage = input->ciImage();
```
```suggestion
    RetainPtr inputImage = input->ciImage();
```

lilyspiniolas commented on Source/WebCore/platform/graphics/coreimage/SourceAlphaCoreImageApplier.mm:
```diff
@@ -0,0 +1,69 @@
+    if (!inputImage)
+        return false;
+
+    auto alphaFilter = [CIFilter filterWithName:@"CIColorMatrix"];
```
```suggestion
    RetainPtr alphaFilter = [CIFilter filterWithName:@"CIColorMatrix"];
```

nullhook commented on Source/WebCore/platform/mac/VideoPresentationInterfaceMac.mm:
```diff
@@ -45,8 +45,7 @@
 
-SOFTLINK_AVKIT_FRAMEWORK()
-SOFT_LINK_CLASS_OPTIONAL(AVKit, AVValueTiming)
+SOFT_LINK_FRAMEWORK_FOR_SOURCE(AVKit, AVValueTiming)
```
@jyavenard changed here.

shallawa commented on Source/WebCore/platform/graphics/coreimage/SourceAlphaCoreImageApplier.mm:
```diff
@@ -0,0 +1,69 @@
+#import "IOSurface.h"
+#import "ImageBuffer.h"
+#import "NativeImage.h"
+#import <CoreImage/CIContext.h>
```
I think these headers are not needed.

shallawa commented on Source/WebCore/SourcesCocoa.txt:
```diff
@@ -500,6 +500,7 @@ platform/graphics/coreimage/FEColorMatrixCoreImageApplier.mm
 platform/graphics/coreimage/FEComponentTransferCoreImageApplier.mm
 platform/graphics/coreimage/FilterImageCoreImage.mm
 platform/graphics/coreimage/SourceGraphicCoreImageApplier.mm
+platform/graphics/coreimage/SourceAlphaCoreImageApplier.mm
```
I think the files names should be sorted in this file.

nullhook committed:
Add Core Image accelerated path for SourceAlpha
https://bugs.webkit.org/show_bug.cgi?id=297018
rdar://problem/157696906

Reviewed by Said Abou-Hallawa.

Implement accelerated rendering for the SourceAlpha
using Core Image's CIColorMatrix

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/coreimage/SourceAlphaCoreImageApplier.h: Added.
* Source/WebCore/platform/graphics/coreimage/SourceAlphaCoreImageApplier.mm: Added.
(WebCore::SourceAlphaCoreImageApplier::apply const):
* Source/WebCore/platform/graphics/filters/SourceAlpha.cpp:
(WebCore::SourceAlpha::supportedFilterRenderingModes const):
(WebCore::SourceAlpha::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/SourceAlpha.h:
* Source/WebCore/platform/mac/VideoPresentationInterfaceMac.mm:

Canonical link: https://commits.webkit.org/298317@main
